### PR TITLE
[Deepseek][Jax] Refactoring deepseek model file

### PR DESCRIPTION
**Pull request by lk-chen**
# Description

With this PR:
* `jrplatin/DeepSeek-R1-1D-Subchannel-256-FP8` is no longer supported
*  it's recommended to use official `deepseek-ai/DeepSeek-R1` checkpoint
* reverts #1810 , otherwise has accuracy drop

# Tests

```
export CURRENT_VLLM_MODEL=deepseek-ai/DeepSeek-R1
VLLM_MLA_DISABLE=1 MOE_REQUANT_BLOCK_SIZE=256 MOE_REQUANT_WEIGHT_DTYPE=fp4 USE_UNFUSED_MEGABLOCKS=0 NEW_MODEL_DESIGN=1 TPU_BACKEND_
TYPE=jax \
MODEL_IMPL_TYPE=flax_nnx vllm serve $CURRENT_VLLM_MODEL \
    --gpu-memory-utilization=0.95 --max-model-len=5120 --max-num-seqs=16 --max-num-batched-tokens 128 \
    --disable-log-requests \
    --no-enable-prefix-caching \
    --tensor-parallel-size 8 \
    --additional_config='{"sharding": {"sharding_strategy": {"enable_dp_attention": true, "expert_parallelism": 1, "tensor_parallel
ism": 8}}, "sparse_matmul": "True"}' \
    --no-async-scheduling
```
```
export CURRENT_VLLM_MODEL=deepseek-ai/DeepSeek-R1
python scripts/vllm/benchmarking/benchmark_serving.py \
    --model $CURRENT_VLLM_MODEL \
    --backend vllm \
    --dataset-name mmlu \
    --dataset-path=/mnt/pd/mmlu/data/test/ \
    --num-prompts 1000 \
    --run_eval
```

MMLU score
```
============ Serving Benchmark Result ============
Successful requests:                     1000
Benchmark duration (s):                  80.84
Total input tokens:                      180362
Total generated tokens:                  2000
Request throughput (req/s):              12.37
Output token throughput (tok/s):         24.74
Total token throughput (tok/s):          2255.73
---------------Time to First Token----------------
Mean TTFT (ms):                          37722.09
Median TTFT (ms):                        37329.85
P99 TTFT (ms):                           79468.72
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          56.53
Median TPOT (ms):                        56.50
P99 TPOT (ms):                           61.18
---------------Inter-token Latency----------------
Mean ITL (ms):                           56.53
Median ITL (ms):                         56.50
P99 ITL (ms):                            61.18
---------------Inter-token Latency----------------
Mean ITL (ms):                           56.53
Median ITL (ms):                         56.50
P99 ITL (ms):                            61.18
----------------End-to-end Latency----------------
Mean E2EL (ms):                          37778.61
Median E2EL (ms):                        37385.90
P99 E2EL (ms):                           79526.16
==================================================
Evaluating MMLU...

Results

{'accuracy': 0.785, 'gen_num': 1000}
```

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/jax/quantization/fp8.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ def load_fp8_weight(jax_param: nnx.Param, torch_weight: torch.Tensor,
         )
         jax_weight = jax_weight.astype(jax_param[...].dtype)
 
-    jax_param[...] = shard_put(jax_weight, spec, mesh=mesh)
```
Comment (jrplatin) [inline tpu_inference/layers/jax/quantization/fp8.py:79]: Typo? I thought we wanted to use [....] moving forward?

File: tpu_inference/models/jax/utils/qwix/qwix_utils.py
Comment (jrplatin) [inline tpu_inference/models/jax/utils/qwix/qwix_utils.py:1]: In the comments, there's a few references to DeepSeek -- can you please remove them since we aren't using Qwix anymore for this model? Thanks!

File: tpu_inference/models/jax/deepseek_v3.py [hunk 2]
```diff
@@ -13,56 +13,54 @@
...
 from jax import lax
-from jax.sharding import Mesh, NamedSharding
+from jax.sharding import Mesh
 from jax.sharding import PartitionSpec as P
 from jaxtyping import Float
 from torchax.ops.mappings import j2t_dtype
 from vllm.config import VllmConfig
 
 from tpu_inference import utils
+from tpu_inference.kernels.mla.v1.kernel import mla_ragged_paged_attention
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py]: Not used? Pretty post #1785 this isn't needed

File: tpu_inference/models/jax/deepseek_v3.py [hunk 3]
```diff
@@ -78,6 +76,48 @@ def _weight_init(random_init: bool):
 
 modeling_flax_utils = FlaxUtils()
 
+num_local_experts: int = 256
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py:73]: Nit: these are constants right? Maybe uppercase them?

File: tpu_inference/models/jax/deepseek_v3.py [hunk 4]
```diff
@@ -676,21 +720,20 @@ def __post_init__(self, rngs: nnx.Rngs):
 
 
 @dataclass(kw_only=True)
-class DeepseekV3MoE(nnx.Module):
+class SharedFusedMoe(JaxMoE):
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py:710]: Should this be in the `deepseek_v3.py` file? Doesn't seem like like this specific to DS?

File: tpu_inference/layers/jax/quantization/fp8.py (hunk 1 above)
Comment (lk-chen) [reply tpu_inference/layers/jax/quantization/fp8.py:79]: Updated to `set_raw_value`

File: tpu_inference/models/jax/utils/qwix/qwix_utils.py
Comment (lk-chen) [reply tpu_inference/models/jax/utils/qwix/qwix_utils.py:1]: I deleted some, and marked some as deprecated. Let's clean up in next PR

File: tpu_inference/models/jax/deepseek_v3.py (hunk 3 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py:73]: Instead of fixing the format, we should read these configs from HF config.

These lines were copied from previous code, I've added TODO above

File: tpu_inference/models/jax/deepseek_v3.py [hunk 5]
```diff
@@ -194,25 +236,25 @@ def __post_init__(self, rngs: nnx.Rngs):
...
-                                        param_dtype=self.dtype,
                                         dtype=self.dtype,
-                                        rngs=rngs,
-                                        quant_config=self.quant_config)
-
-        self.kv_a_layernorm = JaxRmsNorm(self.kv_lora_rank,
-                                         epsilon=self.rms_norm_eps,
-                                         scale_init=nnx.with_partitioning(
-                                             init_fn, (None, )),
-                                         param_dtype=self.dtype,
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py:212]: We are omitting this now on purpose right?

File: tpu_inference/models/jax/deepseek_v3.py [hunk 6]
```diff
@@ -676,21 +720,20 @@ def __post_init__(self, rngs: nnx.Rngs):
...
 
 @dataclass(kw_only=True)
-class DeepseekV3MoE(nnx.Module):
+class SharedFusedMoe(JaxMoE):
     """
-    Corresponds to vLLM's DeepseekV2MoE.
+    Corresponds to vLLM's SharedFusedMoe.
     Handles the routed and shared experts + the relevant forward pass.
 
     Reference here: https://github.com/vllm-project/vllm/blob/2b465570e6dd327e8422ef9c87e9b2b1454ceaed/vllm/model_executor/models/deepseek_v2.py#L223
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py]: Update this ref?

File: tpu_inference/models/jax/deepseek_v3.py [hunk 7]
```diff
@@ -700,23 +743,115 @@ def __call__(self, x_TD: jax.Array) -> jax.Array:
...
+class DeepseekV2Moe(JaxModule):
+
+    def __init__(self,
+                 *,
+                 mesh,
+                 num_expert_parallelism,
+                 moe_backend,
+                 quant_config,
+                 rng,
+                 prefix: str = ""):
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py:747]: Would be good to add a permalink ref to vLLM here

File: tpu_inference/models/jax/deepseek_v3.py [hunk 8]
```diff
@@ -700,23 +737,115 @@ def __call__(self, x_TD: jax.Array) -> jax.Array:
...
-    mlp: nnx.Module | DeepseekV3MoE | DeepseekV3MLP
+    def __init__(
+            self,
+            input_layernorm: JaxRmsNorm,
+            post_attention_layernorm: JaxRmsNorm,
+            self_attn: Union[DeepseekV3Attention, DeepseekV3MLA],
+
+            # MLP can be either the Dense MLP (for first k layers) or SharedFusedMoe
+            # TODO: rename to mlp? custom_module seems needlessly confusing
+            custom_module: nnx.Module | SharedFusedMoe | DeepseekV3MLP,
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py]: This was for Qwix reasons, but agreed -- we should rename this

File: tpu_inference/models/jax/deepseek_v3.py [hunk 9]
```diff
@@ -830,774 +991,67 @@ def __call__(self, x_TD: Float) -> Tuple[Float, Float]:
...
         def _create_deepseek_attention(
-        ) -> Union[DeepseekV3MLA, DeepseekV3Attention]:
+                i: int) -> Union[DeepseekV3MLA, DeepseekV3Attention]:
             if self.use_mla_kernel:
                 query_tnh_spec = P(ShardingAxisName.MLP_TENSOR, None, None)
                 keyvalue_skh_spec = P(ShardingAxisName.MLP_TENSOR, None)
                 attn_o_tnh_spec = P(ShardingAxisName.MLP_TENSOR, None, None)
                 anh_sharding = (None, ShardingAxisName.MLP_TENSOR, None)
             else:
-                query_tnh_spec = P(None, ShardingAxisName.MLP_TENSOR, None)
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py:1598]: Is this change on purpose? Asking because each of the `_spec` variables suffixes seem to imply these are 3D

File: tpu_inference/models/jax/deepseek_v3.py [hunk 10]
```diff
@@ -1691,143 +1145,53 @@ def get_decoder_layer(layer_index: int):
...
-                custom_module = JaxMoE(
-                    dtype=dtype,
-                    num_local_experts=num_local_experts,
-                    apply_expert_weight_before_computation=False,
-                    expert_axis_name=self.expert_axis_name,
-                    num_expert_parallelism=self.num_expert_parallelism,
-                    hidden_size=hidden_size,
-                    intermediate_size_moe=moe_intermediate_size,
-                    num_experts_per_tok=num_experts_per_token,
+                # moe_dtype = jnp.float8_e4m3fn if self.weight_loader.is_native_fp8_model else vllm_config.model_config.hf_config.quantization_config.get(
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py]: Old code?

File: tpu_inference/models/jax/deepseek_v3.py [hunk 11]
```diff
@@ -1836,82 +1200,142 @@ def initialize_cache(self):
...
+
+class DeepseekV3ForCausalLM(JaxModule, LoadableWithIterator):
+
+    def __init__(self, vllm_config: VllmConfig, rng_key: jax.Array,
+                 mesh: Mesh) -> None:
+        if getattr(vllm_config.model_config, "quantization", None) == "fp8":
+            # `get_tpu_quantization_config` returns None for "fp8" because
+            # the work in #1623 is not fully merged. So this block overrides
+            # the logic to return Fp8Config when model_config indicates fp8.
+            # TODO(#1623): Remove this block when `get_tpu_quantization_config`
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py:1236]: Is this still relevant or old?

File: tpu_inference/models/jax/deepseek_v3.py (hunk 4 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py:710]: it's not, but refactored from existing code.
So I'd rather leave it in ds_v3.py in this PR, so people can cross compare before/after more easily.

We can always move it to another file afterwards (when new model coming)

File: tpu_inference/models/jax/deepseek_v3.py [hunk 12]
```diff
@@ -700,23 +737,115 @@ def __call__(self, x_TD: jax.Array) -> jax.Array:
...
+            mesh=mesh,
+            hidden_act=hidden_act,
+            rngs=rng,
+            quant_config=quant_config,
+            activation_ffw_td=moe_activation_ffw_td,
+            activation_ffw_ted=moe_activation_ffw_ted,
+            edf_sharding=moe_edf_sharding,
+            efd_sharding=moe_efd_sharding,
+            moe_backend=moe_backend,
+            qwix_quantized_weight_dtype=None,
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py:812]: Maybe in another PR but we should deprecate this variable

File: tpu_inference/models/jax/deepseek_v3.py (hunk 9 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py:1598]: yes, there's an error saying rank mismatch, which is expecting 2-D spec.

File: tpu_inference/models/jax/deepseek_v3.py [hunk 13]
```diff
@@ -78,6 +76,42 @@ def _weight_init(random_init: bool):
 
 modeling_flax_utils = FlaxUtils()
 
+# TODO: read these configs from HF config.
+num_local_experts: int = 256
+vocab_size: int = 129280
+hidden_size: int = 7168
+# NOTE: this dtype may be implicitly overriden if using to Qwix to load in the quantized weights
```
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py]: nit: remove since no longer using qwix

File: tpu_inference/models/jax/deepseek_v3.py
Comment (jrplatin) [inline tpu_inference/models/jax/deepseek_v3.py]: nit: remove

**Comment by jrplatin** [comment]
Can you please add the relevant E2E testing you did for both MHA/MLA to confirm things work?

Also, do any of the (DeepSeek) unit tests need to be updated?

File: tpu_inference/models/jax/deepseek_v3.py (hunk 12 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py:812]: Agree, it's mostly copied cover. I'll leave it in this PR so people can easily compare before/after

File: tpu_inference/models/jax/deepseek_v3.py (hunk 11 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py:1236]: Will have follow up PR to clean up all "TODO(#1623)"

File: tpu_inference/models/jax/deepseek_v3.py (hunk 10 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py]: Removed

File: tpu_inference/models/jax/deepseek_v3.py (hunk 8 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py]: Done

File: tpu_inference/models/jax/deepseek_v3.py (hunk 6 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py]: Done

File: tpu_inference/models/jax/deepseek_v3.py (hunk 7 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py:747]: Done

File: tpu_inference/models/jax/deepseek_v3.py (hunk 5 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py:212]: Ah, not really. should keep param_dtype, thx for catching

File: tpu_inference/models/jax/deepseek_v3.py (hunk 13 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py]: Done

File: tpu_inference/models/jax/deepseek_v3.py (hunk 2 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py]: Rebased, this change is gone

File: tpu_inference/models/jax/deepseek_v3.py
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py]: Done

File: tpu_inference/models/jax/deepseek_v3.py (hunk 4 above)
Comment (jrplatin) [reply tpu_inference/models/jax/deepseek_v3.py:710]: Sounds good

File: tpu_inference/models/jax/deepseek_v3.py (hunk 9 above)
Comment (jrplatin) [reply tpu_inference/models/jax/deepseek_v3.py:1598]: Hmm, so is `query_tnh` not 3d then? Should we update the suffix?

File: tpu_inference/models/jax/deepseek_v3.py (hunk 3 above)
Comment (jrplatin) [reply tpu_inference/models/jax/deepseek_v3.py:73]: OK yeah that sounds reasonable to me!

File: tpu_inference/models/jax/deepseek_v3.py (hunk 9 above)
Comment (lk-chen) [reply tpu_inference/models/jax/deepseek_v3.py:1598]: will postpone the cleanup post this PR

**Comment by lk-chen** [comment]
> Can you please add the relevant E2E testing you did for both MHA/MLA to confirm things work?
> 
> Also, do any of the (DeepSeek) unit tests need to be updated?

I updated description to include MMLU results. test_qwix is updated to exclude DeepSeek related content
